daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/parsers/Parser.scala

```diff
@@ -227,26 +227,47 @@ abstract class CombinatorParser(override val context: RuntimeData)
   extends Parser
   with CombinatorProcessor
 
-final class SeqCompParser(context: RuntimeData, val childParsers: Vector[Parser])
-  extends CombinatorParser(context) {
+final class SeqCompParser(
+  context: RuntimeData,
+  val childParsers: Vector[Parser],
+) extends CombinatorParser(context) {
   override lazy val runtimeDependencies = Vector()
   override def childProcessors = childParsers
 
   override def nom = "seq"
 
-  val numChildParsers = childParsers.size
+  val assertDiscrimExpressions = childParsers.collect { case ae: AssertExpressionEvaluationParser => ae }
+  val discrimExpressions = assertDiscrimExpressions.filter { _.discrim }
+  val assertExpressions = assertDiscrimExpressions.filterNot { _.discrim }
+  val nonAssertChildren = childParsers.diff(assertExpressions)
```

Comment:
I think `nonAssertChildren` includes discriminator parsers so could discriminators get evaluated twice, once in the first loop and once in the discrimnator loop?

Reply:
There's a base class shared between discriminators and regular asserts. They're both a kind of assertion. This scala code may be using that terminology so nonAssertChildren may also mean no discriminators.

Reply:
@stevedlawrence is right.  It should be

```    val nonAssertChildren = childParsers.diff(assertDiscrimExpressions)```

then there won't be duplicates.  Will change.